### PR TITLE
Fix #4423 - enable coverage on one of the linux runners

### LINKED ISSUE #4423: Enable code coverage on our unit tests
Enhancement Request
--------------
Enable code coverage on our unit tests.

## Detailed Description

Code coverage should help identify portion of the codebase that we do not have unit tests for.

## Possible Implementation

use lcov. `sudo apt install lcov gcovr`

Compile with build flags `--coverage -O0 -g` and link flags `--coverage`. Compile in **Debug** mode.

then something along the lines of 

```bash
cd build/

lcov --zerocounters --directory .

# Run the tests
./Products/openstudio_utilities_tests
./Products/openstudio_model_tests
(etc)

srcdir=$(readlink -e ../)

# CAREFUL! base-directory needs to be an absolute path!
	
# no-external excludes anything that isn't in what's passed as directory or base-directory
lcov --capture --directory . --output-file ./lcov.output --no-external --base-directory $srcdir

# Exclude the build directory
lcov -r lcov.output `pwd`/\* -o ./lcov.output.filtered

# Exclude the test directory if need be (or rather --base-directory should be carefully chosen)
lcov -r lcov.output.filtered $srcdir/test/\* -o ./lcov.output.filtered

genhtml ./lcov.output.filtered -o lcov-html --demangle-cpp --function-coverage --rc genhtml_hi_limit=41 --rc genhtml_med_limit=40

xdg-open lcov-html/index.html &
```

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #ISSUENUMBERHERE (IF THIS IS A DEFECT)

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by tijcolem** [comment]
@jmarrec I wonder if we can view the lines by class that are not being covered? 

for example:  https://ci.commercialbuildings.dev/job/openstudio-incremental-ubuntu-20-04/job/PR-4427/9/cobertura/src_model/ Drilling down by class and seeing the lines of code.

**Comment by jmarrec** [comment]
@tijcolem Yes, by using gcovr's flag `--html-details` instead of `--html`. But that will create like a thousands HTML files instead of a single, so we'd need to output it to another path that is a subfolder.
Uploading the XML to codecov.io is another alternative, It'll provide that too.

**Comment by jmarrec** [comment]
@tijcolem Should we build RelWithDebInfo instead of Debug? https://github.com/NREL/cbci_jenkins_libs/blob/6e4dbca35ff2ace2a5766de317f950e2fad8fe82/vars/openstudio_incremental_code_coverage.groovy#L203

File: CMakeLists.txt [hunk 1]
```diff
@@ -278,7 +278,14 @@ if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
...
 # Enable runtime checking features
-if(CMAKE_COMPILER_IS_GNUCC OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES ".*Clang")
+
+  option(ENABLE_COVERAGE "Enable coverage reporting for gcc/clang" OFF)
+  if(ENABLE_COVERAGE)
+    add_definitions(--coverage -O0 -g)
+    set(SANITIZER_LINKER_FLAGS "${SANITIZER_LINKER_FLAGS} --coverage")
+  endif()
+
```
[the remark this answers is not in the document]
Comment (tijcolem) [reply CMakeLists.txt:288]: Let me know what you find out on size diff between Debug and RelWithDebInfo.   Part of me wonders if these incremental builds should all be built using RelWithDebInfo so we can capture core dumps and debug properly if needed.

**Comment by tijcolem** [comment]
I add the option `--html-details` and adjusted the s3 upload to upload the dir vs single file. I think this now good to drop in. I have this schedule on different CI run so it won't backlog any of the existing runners. It's a DEBUG build now but we could change to RELWITHDEBINFO if we wanted/needed to.

**Comment by jmarrec** [comment]
I saw no significant difference between RelWithDebInfo and Debug (in fact RelWithDebInfo somehow was bigger in size, compilation times were similar)

**Pull request merged**